exploration/contextual-options.md

```diff
@@ -0,0 +1,156 @@
+
+Special options or attributes should not conflict with other option names.
+
+## Constraints
+
+User-defined functions are free to use un-namespaced option names
+even if the function identifier does use a namespace.
+
+The `u` namespace is reserved for future standardization.
+
+No literal syntax is provided for values that are lists or sequences of scalar values.
+
+## Proposed Design
+
+Define the expected values and handling for the following options
+wherever they are used:
+
+- `u:id` — A string value that is included as an `id` or other suitable value
+  in the formatted part(s) for the placeholder,
+  or any other structured formatted results.
+  Ignored when formatting to a string, but could show up in error messages.
+- `u:locale` — A comma-delimited list of BCP 47 language tags,
+  or an implementation-defined list of such tags.
+  The tags are parsed, and they replace the _locale_
+  defined in the _formatting context_ for this expression or markup.
+- `u:dir` — One of the string values `ltr`, `rtl`, or `auto`.
+  Replaces the character directionality
+  defined in the _formatting context_ for this expression or markup.
+
+Error handling should be well defined for invalid values.
+
+Additional restrictions could be imposed,
+e.g. requiring that each `u:id` is unique within a formatted message.
+
+Drop variable values from the `attribute` rule:
+
+```diff
+-attribute      = "@" identifier [[s] "=" [s] (literal / variable)]
++attribute      = "@" identifier [[s] "=" [s] literal]
+```
+
+## Alternatives Considered
+
+### Do nothing
+
+Continue to [caution](https://github.com/unicode-org/message-format-wg/blob/d38ff326d2381b3ef361e996c3431d1b251518d6/spec/syntax.md#attributes)
+function authors and other implementers away from creating function-specific or implementation-specific option values
+for the use cases presented above.
+
+As should be obvious, the current situation is not tenable in the long term, and should be resolved.
+
+### Do not provide any guidance
+
+Do not include in the spec rules or guidance for declaring formatted part identifiers,
+or overriding the message locale or directionality.
+
+This would mean not defining anything for default registry functions either,
+effectively requiring implementation-specific options like `icu:locale`.
+
+Other functions could use their own definitions and handling for similar options,
+such as `locale` or `x:lang`.
+
+Formatted parts for markup would not be able to directly include an identifier.
+
+### Define options for default registry only
+
+Define at least `locale` and `dir` as options for default registry functions,
+with handling internal to each function implementation.
+
+Other functions could use their own definitions and handling for similar options,
+such as `locale` or `x:lang`.
+
+Formatted parts for markup would not be able to directly include an identifier.
+
+### Use expression attributes
```

Comment:
Expression attributes can be used without a function (or an function implied from the type of the argument, which is the same thing).

So we can say: `{$expDate @locale=fr}`, but we can't say `{$expDate u:locale=fr}`, we are forced to say `{$expDate :date u:locale=fr}`
I don't see it as a problem, a bit more typing in some rare cases, but just saying.